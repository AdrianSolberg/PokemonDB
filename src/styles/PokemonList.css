#PokemonListWrapper {
    z-index: 10;
}

.list {
    list-style-type: none;
    padding-left: 0;
    width: 200px;
    margin: 12px;
    background-color: #303030;
    border-radius: 8px;
    box-shadow: 0 15px 30px #101010;
    max-height: calc(100dvh - 2 * 12px);
    overflow-y: auto;
    transition: max-height 0.3s ease-out;
}

.list li {
    display: flex;
    align-items: center;
    padding: 12px 16px;
    cursor: pointer;
    border-bottom: 1px solid #434343;
    font-weight: bold;
}

.list li:hover {
    background-color: #4a4a4a;
}

.list li.selected {
    background: linear-gradient(
        to bottom,
        rgb(254, 235, 235) 0%,
        rgb(255, 121, 121) 100%
    );
    color: #213547;
}

.list li:last-child {
    border-bottom: none;
}

.list li:first-child.selected {
    border-top-left-radius: 8px;
}

.list li:last-child.selected {
    border-bottom-left-radius: 8px;
}

.list li:first-child:hover {
    border-top-left-radius: 8px;
}

.list li:last-child:hover {
    border-bottom-left-radius: 8px;
}

.list img {
    width: 40px;
    height: 40px;
    margin-right: 10px;
    border-radius: 5px;
}

.list img.refresh {
    width: 30px;
    height: 30px;
}

.list img.loading {
    animation: spin 0.25s linear infinite;
}

.list li:hover > img.refresh {
    animation: spin 0.25s linear;
}

@keyframes spin {
    0% {
        transform: rotate(0deg);
    }
    100% {
        transform: rotate(360deg);
    }
}

.hamburger {
    display: none;
    margin: 12px;
    width: 30px;
    height: 30px;
    cursor: pointer;
    flex-direction: column;
    justify-content: space-around;
    align-items: center;
    background: none;
    border: none;
    padding: 0;
}

.hamburger:focus {
    outline: none;
}

.hamburger .hamburgerLine {
    width: 100%;
    height: 4px;
    background-color: #616161;
    border-radius: 5px;
    transition:
        transform 0.3s,
        opacity 0.3s;
}

@media (max-width: 945px) {
    #PokemonListWrapper {
        position: absolute;
        left: 0;
        top: 0;
    }

    .list {
        max-height: 0;
        transition: max-height 0.3s ease-out;
    }

    .list.active {
        /* Icon is 30px. There are 3 margins on 12px each */
        max-height: calc(100dvh - 30px - 3 * 12px);
    }

    .hamburger {
        display: flex;
    }
}

.hamburger.active .hamburgerLine:nth-child(1) {
    transform: translateY(10px) rotate(45deg) scale(1.1);
}
.hamburger.active .hamburgerLine:nth-child(2) {
    opacity: 0;
}
.hamburger.active .hamburgerLine:nth-child(3) {
    transform: translateY(-10px) rotate(-45deg) scale(1.1);
}

.list::-webkit-scrollbar {
    width: 7px;
}

.list:hover::-webkit-scrollbar {
    width: 11px;
}

.list::-webkit-scrollbar-track {
    background: #303030;
    border-top-right-radius: 5px;
    border-bottom-right-radius: 5px;
}

.list:hover::-webkit-scrollbar-track {
    background: #202020;
}

.list::-webkit-scrollbar-thumb {
    background: #7c7c7c;
    border-radius: 10px;
    border: 2px solid #303030;
}

.list:hover::-webkit-scrollbar-thumb {
    background: #7c7c7c;
    border: 3px solid #202020;
}

.list::-webkit-scrollbar-thumb:hover {
    background: #999999;
}
