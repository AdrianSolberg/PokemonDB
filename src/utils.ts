import { PokeAPI } from "pokeapi-types";

export function capitalizeFirstLetter(string: string): string {
    return string.charAt(0).toUpperCase() + string.slice(1);
}

function sortByNameAsc(p1: PokeAPI.Pokemon, p2: PokeAPI.Pokemon): number {
    return p1.name.localeCompare(p2.name);
}

function sortByNameDesc(p1: PokeAPI.Pokemon, p2: PokeAPI.Pokemon): number {
    return p2.name.localeCompare(p1.name);
}

function sortById(p1: PokeAPI.Pokemon, p2: PokeAPI.Pokemon): number {
    return p1.id - p2.id;
}

export const sortingMap: Map<
    string,
    (p1: PokeAPI.Pokemon, p2: PokeAPI.Pokemon) => number
> = new Map<string, (p1: PokeAPI.Pokemon, p2: PokeAPI.Pokemon) => number>([
    ["name_asc", sortByNameAsc],
    ["name_desc", sortByNameDesc],
    ["id", sortById],
]);

const pokemonTypeColors: Map<string, string> = new Map<string, string>([
    ["normal", "#9FA19F"],
    ["fire", "#E62829"],
    ["water", "#2980EF"],
    ["electric", "#FAC000"],
    ["grass", "#3FA129"],
    ["ice", "#3DCEF3"],
    ["fighting", "#FF8000"],
    ["poison", "#9141CB"],
    ["ground", "#915121"],
    ["flying", "#81B9EF"],
    ["psychic", "#EF4179"],
    ["bug", "#91A119"],
    ["ghost", "#704170"],
    ["dragon", "#5060E1"],
    ["dark", "#624D4E"],
    ["steel", "#60A1B8"],
    ["fairy", "#EF70EF"],
]);
export default pokemonTypeColors;

export function filterAndSortPokemon(
    pokemon: PokeAPI.Pokemon[],
    filters: string[],
    sortingOrder: string
): PokeAPI.Pokemon[] {
    return pokemon
        .filter((p: PokeAPI.Pokemon) => {
            const types: string[] = p.types.map((t) => t.type.name);
            for (let i = 0; i < filters.length; i++) {
                if (!types.includes(filters[i])) return false;
            }
            return true;
        })
        .sort(sortingMap.get(sortingOrder));
}
