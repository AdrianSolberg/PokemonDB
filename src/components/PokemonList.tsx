import { useState } from "react";
import { PokeAPI } from "pokeapi-types";
import "../styles/PokemonList.css";
import { Requests } from "../api/Requests";
import { useQuery } from "@tanstack/react-query";
import { capitalizeFirstLetter } from "../utils";

const PokemonList = ({
    limit,
    offset,
    onPokemonSelect,
}: {
    limit: number;
    offset: number;
    onPokemonSelect: (id: number) => void;
}) => {
    const [selectedIndex, setSelectedIndex] = useState(1);
    const [isMenuOpen, setIsMenuOpen] = useState(false);

    const handleItemClick = (index: number) => {
        setSelectedIndex(index);
        onPokemonSelect(index);
        if (window.innerWidth <= 900) {
            toggleMenu();
        }
    };

    const toggleMenu = () => {
        setIsMenuOpen(!isMenuOpen);
    };

    function fetchData(): Promise<PokeAPI.Pokemon[]> {
        return Requests.getPokemonSliceAllData(limit, offset);
    }

    const { data, error, isLoading } = useQuery({
        queryKey: ["pokemon", "allData", limit, offset],
        queryFn: fetchData,
    });

    if (isLoading) return <div>Loading...</div>;
    if (error || data == undefined)
        return <div>Error fetching Pok√©mon data</div>;

    return (
        <>
            <button
                className={`hamburger ${isMenuOpen ? "active" : ""}`}
                aria-label="Toggle Menu"
                onClick={toggleMenu}
            >
                <div className="hamburgerLine"></div>
                <div className="hamburgerLine"></div>
                <div className="hamburgerLine"></div>
            </button>

            <ul className={`list ${isMenuOpen ? "active" : ""}`}>
                {data.map((pokemon) => (
                    <li
                        key={pokemon.id}
                        className={
                            selectedIndex == pokemon.id ? "selected" : ""
                        }
                        onClick={() => handleItemClick(pokemon.id)}
                        role="button"
                        aria-pressed={selectedIndex == pokemon.id}
                        tabIndex={0}
                        onKeyDown={(e) => {
                            if (e.key == "Enter" || e.key == " ") {
                                handleItemClick(pokemon.id);
                            }
                        }}
                    >
                        <img
                            src={pokemon.sprites.front_default}
                            alt={`${capitalizeFirstLetter(pokemon.name)} image`}
                        />
                        {capitalizeFirstLetter(pokemon.name)}
                    </li>
                ))}
            </ul>
        </>
    );
};

export default PokemonList;
