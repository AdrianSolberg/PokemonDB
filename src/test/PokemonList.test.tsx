import { describe, it, expect, vi, Mock } from "vitest";
import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import PokemonList from "../components/PokemonList";
import { useInfiniteQuery } from "@tanstack/react-query";
import "@testing-library/jest-dom";
import mockPokemonData from "./mocks/mockPokemonData.json";
import { useAppContext } from "../utils";

vi.mock(import("../utils"), async (importOriginal) => {
    const actual = await importOriginal();
    return {
        ...actual,
        useAppContext: vi.fn(),
    };
});

vi.mock(import("@tanstack/react-query"), async (importOriginal) => {
    const actual = await importOriginal();
    return {
        ...actual,
        useInfiniteQuery: vi.fn(),
    };
});

describe("PokemonList - General tests", () => {
    const changeSelectedPokemonId = vi.fn();

    beforeEach(() => {
        (useAppContext as Mock).mockReturnValue({
            selectedPokemonId: 1,
            changeSelectedPokemonId,
            sortingOrder: "asc",
            filters: {},
            favorites: [],
            currentPokemonList: mockPokemonData,
            setCurrentPokemonList: vi.fn(),
        });
    });

    it("should display loading state", () => {
        (useInfiniteQuery as Mock).mockReturnValue({
            isLoading: true,
            data: undefined,
            error: null,
        });
        render(<PokemonList limit={10} />);
        expect(screen.getByText("Loading...")).toBeInTheDocument();
    });

    it("should display error", () => {
        (useInfiniteQuery as Mock).mockReturnValue({
            isLoading: false,
            data: undefined,
            error: new Error("This is an error"),
        });
        render(<PokemonList limit={10} />);
        expect(
            screen.getByText("Error fetching Pokémon data")
        ).toBeInTheDocument();
    });

    it("should display Pokémon list", async () => {
        (useInfiniteQuery as Mock).mockReturnValue({
            isLoading: false,
            data: { pages: [mockPokemonData] },
            error: null,
            hasNextPage: false,
        });

        // mock the scrollIntoView function as it fails the test and it is not necessary for the test.
        window.HTMLLIElement.prototype.scrollIntoView = vi.fn();
        render(<PokemonList limit={10} />);

        expect(await screen.findByText("Bulbasaur")).toBeInTheDocument();
        expect(await screen.findByText("Ivysaur")).toBeInTheDocument();
    });

    it("should load more Pokémon when 'Load More' is clicked", async () => {
        const fetchNextPageMock = vi.fn();
        (useInfiniteQuery as Mock).mockReturnValue({
            isLoading: false,
            data: { pages: [mockPokemonData] },
            error: null,
            hasNextPage: true,
            fetchNextPage: fetchNextPageMock,
            isFetchingNextPage: false,
        });

        render(<PokemonList limit={10} />);
        const loadMoreButton = await screen.findByText("Load More");

        userEvent.click(loadMoreButton);
        await waitFor(() => expect(fetchNextPageMock).toHaveBeenCalled());
    });

    it("should select a Pokemon on click", async () => {
        render(<PokemonList limit={10} />);

        userEvent.click(await screen.findByText("Ivysaur"));

        await waitFor(() => {
            expect(changeSelectedPokemonId).toHaveBeenCalledWith(2);
        });
    });
});
