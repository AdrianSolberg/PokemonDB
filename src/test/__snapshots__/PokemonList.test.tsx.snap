// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PokemonList - Snapshot test > matches snapshot 1`] = `
<DocumentFragment>
  <div
    id="PokemonListWrapper"
  >
    <button
      aria-label="Toggle Menu"
      class="hamburger "
    >
      <div
        class="hamburgerLine"
      />
      <div
        class="hamburgerLine"
      />
      <div
        class="hamburgerLine"
      />
    </button>
    <ul
      class="list "
    >
      <li
        aria-pressed="true"
        class="selected"
        role="button"
        tabindex="0"
      >
        <img
          alt="Bulbasaur image"
          src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png"
        />
        Bulbasaur
      </li>
      <li
        aria-pressed="false"
        class=""
        role="button"
        tabindex="0"
      >
        <img
          alt="Ivysaur image"
          src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/2.png"
        />
        Ivysaur
      </li>
      <li
        aria-pressed="false"
        class=""
        role="button"
        tabindex="0"
      >
        <img
          alt="Charmander image"
          src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/4.png"
        />
        Charmander
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
